src/providers/core_provider/mod.rs

```diff
@@ -75,16 +66,42 @@ impl Provide for CoreProvider {
 pub struct CoreProviderBuilder {
     version_maj: Option<u8>,
     version_min: Option<u8>,
-    providers: Option<Vec<ProviderInfo>>,
+    providers: HashMap<ProviderInfo, HashSet<Opcode>>,
 }
 
 impl CoreProviderBuilder {
-    pub fn new() -> Self {
-        CoreProviderBuilder {
+    pub fn new() -> std::io::Result<Self> {
+        let crate_version: Version = Version::from_str(version!()).or_else(|e| {
+            error!("Error parsing the crate version: {}.", e);
+            Err(Error::new(
+                ErrorKind::InvalidData,
+                "crate version number has invalid format",
+            ))
+        })?;
+
+        let mut providers = HashMap::new();
+        let _ = providers.insert(
+            ProviderInfo {
+                // Assigned UUID for this provider: 47049873-2a43-4845-9d72-831eab668784
+                uuid: Uuid::parse_str("47049873-2a43-4845-9d72-831eab668784").or_else(|_| Err(Error::new(
+                    ErrorKind::InvalidData,
+                    "provider UUID is invalid",
+                )))?,
+                description: String::from("Software provider that implements only administrative (i.e. no cryptographic) operations"),
+                vendor: String::new(),
+                version_maj: crate_version.major,
+                version_min: crate_version.minor,
+                version_rev: crate_version.patch,
+                id: ProviderID::Core,
+            },
+            SUPPORTED_OPCODES.iter().copied().collect(),
+        );
```

Comment:
I think having this in an independant method (`describe`) is quite nice to identify this information more easily and mirror the other providers.

Reply:
Yeah, I ended up shuffling things around a few times, in the current format it makes sense to re-implement describe on the core provider

Reply:
Oh, yeah! `describe` requires that an instance of the object already exists, so then you can't add it before you build it. You could build the provider, describe it and then add that information in itself, but that feels counter-intuitive as well. I think I'll leave it like this.